crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,957 @@
+    /// might take one round trip to take effect.
+    pub fn unsubscribe(&self, room_id: OwnedRoomId) {
+        if self.subscriptions.lock_mut().remove(&room_id).is_some() {
+            self.unsubscribe.lock_mut().push_cloned(room_id);
+        }
+    }
+
+    /// Lookup a specific room
+    pub fn get_room(&self, room_id: OwnedRoomId) -> Option<SlidingSyncRoom> {
+        self.rooms.lock_ref().get(&room_id).cloned()
+    }
+
+    /// Lookup a set of rooms
+    pub fn get_rooms<I: Iterator<Item = OwnedRoomId>>(
+        &self,
+        room_ids: I,
+    ) -> Vec<Option<SlidingSyncRoom>> {
+        let rooms = self.rooms.lock_ref();
+        room_ids.map(|room_id| rooms.get(&room_id).cloned()).collect()
+    }
+
+    async fn handle_response(
+        &self,
+        resp: v4::Response,
+        views: &[SlidingSyncView],
+    ) -> anyhow::Result<UpdateSummary> {
+        self.client.process_sliding_sync(resp.clone()).await?;
+        tracing::info!("main client processed.");
+        self.pos.replace(Some(resp.pos));
+        let mut updated_views = Vec::new();
+        if resp.lists.len() != views.len() {
+            bail!("Received response for {} lists, yet we have {}", resp.lists.len(), views.len());
+        }
+
+        for (view, updates) in views.iter().zip(resp.lists.iter()) {
+            let count: u32 =
+                updates.count.try_into().expect("the list total count convertible into u32");
+            tracing::trace!("view {:?}  update: {:?}", view.name, !updates.ops.is_empty());
+            if !updates.ops.is_empty() && view.handle_response(count, &updates.ops)? {
+                updated_views.push(view.name.clone());
+            }
+        }
+
+        let mut rooms = Vec::new();
+        let mut rooms_map = self.rooms.lock_mut();
+        for (id, room_data) in resp.rooms.iter() {
+            if let Some(mut r) = rooms_map.remove(id) {
+                r.update(room_data);
+                rooms_map.insert_cloned(id.clone(), r);
+                rooms.push(id.clone());
+            } else {
+                rooms_map.insert_cloned(
+                    id.clone(),
+                    SlidingSyncRoom::from(id.clone(), room_data.clone()),
+                );
+                rooms.push(id.clone());
+            }
+        }
+
+        Ok(UpdateSummary { views: updated_views, rooms })
+    }
+
+    /// Create the inner stream for the view.
+    ///
+    /// Run this stream to receive new updates from the server.
+    pub async fn stream<'a>(
+        &self,
+    ) -> anyhow::Result<impl Stream<Item = anyhow::Result<UpdateSummary>> + '_> {
+        let views = self.views.lock_ref().to_vec();
+        let _pos = self.pos.clone();
+
+        // FIXME: hack for while the sliding sync server is on a proxy
+        let client = self.client.clone();
+        let req_config =
+            self.homeserver.as_ref().map(|hs| RequestConfig::default().homeserver(hs.to_string()));
```

Comment:
So the sliding-sync proxy can't be used as the main / only HS url, it must only be used for the sliding-sync endpoint? :S

Reply:
> main / only HS url, it must only be used for the sliding-sync endpoint?

yes. Despite the name, it doesn't proxy the other requests. So, unfortunately, for everything but the new sync-api, we need to fallback to the actual home server. It's how it is, not my choice...

Reply:
Alright. Can we maybe split out a few internal functions with extra homeserver parameters then? I really don't like that the consequence of this is a change of public API (`RequestConfig` no longer being `Copy`).